components/concept/awesomebar/src/main/java/mozilla/components/concept/awesomebar/AwesomeBar.kt

```diff
@@ -108,31 +132,29 @@ interface AwesomeBar {
         /**
          * Fired whenever the user changes their input, after they have started interacting with the awesome bar.
          *
-         * This is a synchronous call. It is up to the [AwesomeBar] implementation to perform this operation on a
-         * non-main thread. For example the `browser-awesomebar` component wrap this call in an async block inside a
-         * coroutine:
-         *
-         * ```Kotlin
-         * jobs = providers.map { provider ->
-         *     launch(UI) {
-         *        val suggestions = async(dispatcher) { provider.onInputChanged(text) }
-         *        // Do something with suggestions
-         *     }
-         * }
-         * ```
+         * This is a suspending function. An [AwesomeBar] implementation is expected to invoke this method from a
+         * [Coroutine](https://kotlinlang.org/docs/reference/coroutines-overview.html). This allows the [AwesomeBar]
+         * implementation to group and cancel calls to multiple providers.
          *
-         * This allows the awesome bar component to be in control of the jobs launched and lets it cancel jobs as soon
-         * as they are no longer needed. For the provider this makes the implementation easier as it does not need to
-         * deal with asynchronous cancellable tasks.
+         * Coroutine cancellation is cooperative. A coroutine code has to cooperate to be cancellable:
+         * https://github.com/Kotlin/kotlinx.coroutines/blob/master/docs/cancellation-and-timeouts.md
          *
          * @param text The current user input in the toolbar.
          * @return A list of suggestions to be displayed by the [AwesomeBar].
          */
-        fun onInputChanged(text: String): List<Suggestion>
+        suspend fun onInputChanged(text: String): List<Suggestion>
```

Comment:
Why not do something like `Deferred<List<Suggestion>>`? `await` is a suspending function, which allows for the same usage pattern. It kind of matches the intent a bit better, too? This is still new to me, so it'd be great to see why you chose plain `suspend`.

One reason I can think of is that the implementation wouldn't be forced to use `CompleteableDeferred` or something similar if it's non-blocking (like an in-memory thing..).

Reply:
Deferred means the provider will be responsible for threading. That's not what I want. The caller will wrap that in async *in its own scope* so that it can shutdown everything once it is done.